vortex-layout/src/layouts/dict/mod.rs

```diff
@@ -1,43 +1,139 @@
+use vortex_dtype::{DType, FieldMask, PType};
+use vortex_error::{VortexExpect, VortexResult, vortex_bail, vortex_panic};
+
+use crate::children::LayoutChildren;
+use crate::segments::{SegmentId, SegmentSource};
+use crate::{
+    LayoutChildType, LayoutEncodingRef, LayoutId, LayoutReaderRef, LayoutRef, VTable, vtable,
+};
 
-use crate::segments::SegmentSource;
-use crate::{DICT_LAYOUT_ID, Layout, LayoutReader, LayoutReaderExt as _, LayoutVTable};
+vtable!(Dict);
 
-#[derive(Default, Debug)]
-pub struct DictLayout;
+impl VTable for DictVTable {
+    type Layout = DictLayout;
+    type Encoding = DictLayoutEncoding;
+    type Metadata = ProstMetadata<DictLayoutMetadata>;
 
-impl LayoutVTable for DictLayout {
-    fn id(&self) -> crate::LayoutId {
-        DICT_LAYOUT_ID
+    fn id(_encoding: &Self::Encoding) -> LayoutId {
+        LayoutId::new_ref("vortex.dict")
     }
 
-    fn reader(
-        &self,
-        layout: Layout,
-        segment_source: &Arc<dyn SegmentSource>,
-        ctx: &ArrayContext,
-    ) -> VortexResult<Arc<dyn LayoutReader>> {
-        Ok(DictReader::try_new(layout, segment_source, ctx)?.into_arc())
+    fn encoding(_layout: &Self::Layout) -> LayoutEncodingRef {
+        LayoutEncodingRef::new_ref(DictLayoutEncoding.as_ref())
+    }
+
+    fn row_count(layout: &Self::Layout) -> u64 {
+        layout.codes.row_count()
+    }
+
+    fn dtype(layout: &Self::Layout) -> &DType {
+        layout.values.dtype()
+    }
+
+    fn metadata(layout: &Self::Layout) -> Self::Metadata {
+        ProstMetadata(DictLayoutMetadata::new(
+            PType::try_from(layout.codes.dtype()).vortex_expect("ptype"),
+        ))
+    }
+
+    fn segment_ids(_layout: &Self::Layout) -> Vec<SegmentId> {
+        vec![]
+    }
+
+    fn nchildren(_layout: &Self::Layout) -> usize {
+        2
+    }
+
+    fn child(layout: &Self::Layout, idx: usize) -> VortexResult<LayoutRef> {
+        match idx {
+            0 => Ok(layout.values.clone()),
+            1 => Ok(layout.codes.clone()),
+            _ => vortex_bail!("Unreachable child index: {}", idx),
+        }
+    }
+
+    fn child_type(_layout: &Self::Layout, idx: usize) -> LayoutChildType {
+        match idx {
+            0 => LayoutChildType::Auxiliary("values".into()),
+            1 => LayoutChildType::Transparent("codes".into()),
+            _ => vortex_panic!("Unreachable child index: {}", idx),
+        }
     }
 
     fn register_splits(
-        &self,
-        layout: &Layout,
+        layout: &Self::Layout,
         field_mask: &[FieldMask],
         row_offset: u64,
         splits: &mut BTreeSet<u64>,
     ) -> VortexResult<()> {
-        layout
-            .child(1, layout.dtype().clone(), "codes")?
-            .register_splits(field_mask, row_offset, splits)
+        layout.codes.register_splits(field_mask, row_offset, splits)
+    }
+
+    fn new_reader(
+        layout: &Self::Layout,
+        name: &Arc<str>,
```

Comment:
How about we make this ArcRef<str>? Though that's likely minor. Could also create a `const UNNAMED: Arc<str> = Arc::from("");` and clone that instead of creating new ones